### PR TITLE
Add Integration specs

### PR DESCRIPTION
**Pull request by Achilles-Dev**
## Changes in this PR:
- [ ] Fixed all the N+1 problems encountered when fetching all posts and their comments for a user.
- [ ] Setup Capybara to write integration tests for all views in the project.
- Login page:
  - I can see the username and password inputs and the "Submit" button.
  - When I click the submit button without filling in the username and the password, I get a detailed error.
  - When I click the submit button after filling in the username and the password with incorrect data, I get a detailed 
  error.
  - When I click the submit button after filling in the username and the password with correct data, I am redirected to the root page.
- User index page:
  - I can see the username of all other users.
  - I can see the profile picture for each user.
  - I can see the number of posts each user has written.
  - When I click on a user, I am redirected to that user's show page.
- User show page:
  - I can see the user's profile picture.
  - I can see the user's username.
  - I can see the number of posts the user has written.
  - I can see the user's bio.
  - I can see the user's first 3 posts.
  - I can see a button that lets me view all of a user's posts.
  - When I click a user's post, it redirects me to that post's show page.
  - When I click to see all posts, it redirects me to the user's post's index page.
- User post index page:
  - I can see the user's profile picture.
  - I can see the user's username.
  - I can see the number of posts the user has written.
  - I can see a post's title.
  - I can see some of the post's body.
  - I can see the first comments on a post.
  - I can see how many comments a post has.
  - I can see how many likes a post has.
  - I can see a section for pagination if there are more posts than fit on the view.
  - When I click on a post, it redirects me to that post's show page. 
- Post show page:
  - I can see the post's title.
  - I can see who wrote the post.
  - I can see how many comments it has.
  - I can see how many likes it has.
  - I can see the post body.
  - I can see the username of each commenter.
  - I can see the comment each commenter left.

### CODE REVIEW

File: spec/views/user_show_page_spec.rb [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+    visit user_path(@user1)
+    click_link 'Post 4'
+    expect(current_path).to eq user_post_path(@user1, @post4)
+  end
+
+  it "click on Oykek to redirect to user's post's index page" do
+    visit user_path(@user1)
+    click_link 'Oybek'
+    expect(current_path).to eq user_posts_path(@user1)
+  end
```
Comment (medaminedev66) [inline spec/views/user_show_page_spec.rb:58]: Please ensure the User **show page** has the following tests:
- Tests that redirects to the user's post's index page when you click on a **see all posts**.

File: spec/views/posts_index_spec.rb [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+  it 'shows the first comment on a post' do
+    Comment.create(text: 'first comment', user_id: @user.id, post_id: Post.first.id)
+    visit(user_posts_path(@user.id))
+    expect(page).to have_content Comment.first.text
+  end
+
+  it 'I should see how many likes a post has' do
+    expect(page).to have_content 'Likes: 0'
+  end
+end
```
Comment (medaminedev66) [inline spec/views/posts_index_spec.rb:60]: Please ensure the User **post show page** has the following tests: 
- Tests that asserts you can see how many comments a post has.
- Tests that asserts you can see the number of posts the user has written.
- Tests that asserts you can see a section for pagination if there are more posts than fit on the view.
- Tests that redirects to the post's show page when you click on a post.

File: spec/views/posts_show_spec.rb [hunk 3]
```diff
@@ -0,0 +1,52 @@
+require 'rails_helper'
+
```
Comment (medaminedev66) [inline spec/views/posts_show_spec.rb:2]: - [ ] There are some tests that fail when I run `rspec spec/views/`. Please ensure all the tests are passed successfully.

image

**Review by medaminedev66 — CHANGES_REQUESTED** [review]
Hi @Achilles-Dev 😃, :wave:, 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!

## Highlights
- The repo looks professional
- The app solves N+1 problem when fetching all posts and their comments for a user
- The app has integration tests

## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._
Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: spec/views/posts_index_spec.rb (hunk 2 above)
Comment (Achilles-Dev) [reply spec/views/posts_index_spec.rb:60]: I believe some of these tests such as "_**Tests that assert you can see the number of posts the user has written."**_ are for the **post index page**. I have implemented them there

File: spec/views/posts_index_spec.rb [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+    (1..5).each do |n|
+      post = Post.create(title: "First post #{n}", text: 'Post content', user_id: @user.id, comments_counter: 0,
+                         likes_counter: 0)
+      post.counter_updater
+    end
+    fill_in 'user[email]', with: 'sam@gmail.com'
+    fill_in 'user[password]', with: 'sam'
+    visit(user_posts_path(@user.id))
+    @post = @user.posts.last
+  end
```
Comment (julie-ify) [inline spec/views/posts_index_spec.rb:18]: - [ ] Most of your tests still failed, which was pointed out by the previous code reviewer. From the failure message on the terminal, the problem comes from when you have something like this;

```ruby
 def new
    like = Like.new
    respond_to do |format|
      format.html { render :new, locals: { like: } }
    end
  end
```

But please modify all the places you have something like the above code snippet to have a different variable name so you can pass it as a key-value pair 👇 to see if your tests pass.

```ruby
 def new
    new_like = Like.new
    respond_to do |format|
      format.html { render :new, locals: { like: new_like } }
    end
  end
```

Code_9PsbxnmvuW

**Review by julie-ify — CHANGES_REQUESTED** [review]
Hi @Achilles-Dev, 

Great work on making the changes requested by a previous reviewer 👏 
You've done well implementing some of the requested changes, but there are still some which aren't addressed yet.


 ## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: spec/views/posts_index_spec.rb (hunk 4 above)
Comment (Achilles-Dev) [reply spec/views/posts_index_spec.rb:18]: Thank you. I saw that comment from the last reviewer, however, whenever I run the specs, I do not see those errors which was why I could not fix it. I have removed the part that was causing those errors from your suggestion. 
specs

**Review by ShadyShawkat — APPROVED** [review]
## Hi @Achilles-Dev, 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

## To Highlight
- All tests are passing ✔️ 
- Correct use of GitFlow ✔️ 
- Descriptive commit messages ✔️ 
- Your code is clean and well documented ✔️

Image

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**